src/envoy/alts/tsi_transport_socket.cc

```diff
@@ -0,0 +1,278 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "src/envoy/alts/tsi_transport_socket.h"
+
+#include "common/common/assert.h"
+#include "common/common/enum_to_int.h"
+#include "envoy/event/dispatcher.h"
+
+namespace Envoy {
+namespace Security {
+
+TsiSocket::TsiSocket(HandshakerFactoryCb handshaker_cb)
+    : handshaker_cb_(handshaker_cb), raw_buffer_callbacks_(*this) {
+  raw_buffer_socket_.setTransportSocketCallbacks(raw_buffer_callbacks_);
+}
+
+TsiSocket::~TsiSocket() {
+  ASSERT(!handshaker_);
+  tsi_frame_protector_destroy(frame_protector_);
+  frame_protector_ = nullptr;
+}
+
+void TsiSocket::setTransportSocketCallbacks(
+    Envoy::Network::TransportSocketCallbacks &callbacks) {
+  callbacks_ = &callbacks;
+
+  handshaker_ = handshaker_cb_(callbacks.connection().dispatcher());
+  handshaker_->setHandshakerCallbacks(*this);
+}
+
+std::string TsiSocket::protocol() const { return ""; }
```

Comment:
why protocol is empty?

Reply:
ALTS has no knowledge whether the protocol above it is right now, so this is empty string. When ALTS has similar support of ALPN or NPN in its metadata, this will be populated from there.